File: solutions/haskell/157-middle-of-the-linked-list.hs
{-
  Problem 157: Middle of the Linked List (LeetCode 876)
  Difficulty: Easy
  Language: Haskell
-}
data ListNode = ListNode { val :: Int, next :: Maybe ListNode }

middleNode :: Maybe ListNode -> Maybe ListNode
middleNode head =
  let nodes = collect head
      collect Nothing = []
      collect (Just n) = n : collect (next n)
      len = length nodes
  in if null nodes then Nothing
     else Just (nodes !! (len `div` 2))

File: solutions/elixir/157-middle-of-the-linked-list.ex
# Problem 157: Middle of the Linked List (LeetCode 876)
# Difficulty: Easy
# Language: Elixir
# 
defmodule MiddleNode do
  def middle(nil), do: nil
  def middle(head) do
    nodes = collect(head)
    Enum.at(nodes, div(length(nodes), 2))
  end
  defp collect(nil), do: []
  defp collect(%{val: _, next: nxt} = node), do: [node | collect(nxt)]
end
